### PR TITLE
Create a new client creation API with functional options.

### PR DESCRIPTION
**Pull request by bogosj**
Proposed v1.0 client.New interface. This will provide for future functional options to further configure the client. We should remove the other New* functions before tagging v1.0.

### CODE REVIEW

**Comment by bogosj** [comment]
I'd also like feedback from @andig

File: client.go [hunk 1]
```diff
@@ -25,9 +25,62 @@ type Client struct {
...
+		return func(c *Client) error {
+			return err
+		}
+	}
+	return WithToken(t)
+}
+
+// New creates a new Tesla API client. You must provided one of WithToken or WithTokenFile
+// functional options to initialize the client with an OAuth token.
+func New(ctx context.Context, options ...func(*Client) error) (*Client, error) {
```
Comment (michaelharo) [inline client.go:53]: This function returns a client so I wonder if it should be called `NewClient`.  I know that currently conflicts with the old `NewClient` but perhaps we don't need to keep it with this update?

File: client.go [hunk 2]
```diff
@@ -25,9 +25,62 @@ type Client struct {
...
+	config := &oauth2.Config{
```
Comment (michaelharo) [inline client.go:54]: outside the scope of this change, but I'm expecting at some point in the future we'll want a WithOAuth2Config option that allows setting a different config.

File: client.go (hunk 1 above)
Comment (michaelharo) [inline client.go:53]: I wonder if having `func(*Client) error` have a defined type would make this cleaner.

Maybe something like
```
type ClientOption func(c *Client) error
```

Then we'd have
```
func WithToken(t *oauth2.Token) ClientOption {
   ...
}
func WithTokenFile(path string) ClientOption {
   ...
}
func New(ctx context.Context, options ...ClientOption) (*Client, error) {
   ...
}
```

Thoughts?

**Comment by andig** [comment]
+1, nice interface.

/cc @uhthomas 

OT: would still be great to get the authorisation stuff right before tagging 1.0

File: client.go (hunk 1 above)
Comment (bogosj) [reply client.go:53]: Agreed, I was going off the convention of "if there's only one constructor, call it New" but it's not in a `client` package so it's not intuitive `tesla.New` is creating a client. Addressed by just removing the deprecated New* methods.

File: client.go (hunk 2 above)
Comment (bogosj) [reply client.go:54]: https://github.com/bogosj/tesla/issues/31

File: client.go (hunk 1 above)
Comment (bogosj) [reply client.go:53]: I like it.

**Pull request merged**